cpp/include/cudf/io_types.h

```diff
@@ -23,27 +23,47 @@
  */
 #pragma once
 
+typedef enum 
+{
+  FILE_PATH,
+  BUFFER,
+  URL
+} resource_type;
+
+typedef struct 
+{  
+  char			*data;			/**< pointer to a host memory buffer	*/
+  size_t		size;			/**< size of the buffer, in bytes		*/
+} buffer_desc;
+
+typedef struct
+{
+  resource_type type;
+
+  const char	*path;			/**< file location						*/
+  buffer_desc	buffer;			/**< host memory buffer					*/
```

Comment:
Having it as a separate resource struct seems unnecessary.
Also, simpler to remove buffer_desc and just have:

```c++
  resource_type type;			/**< Type of source of CSV data */
  const char	*file_or_buffer;	/**< If type is FILE_PATH, read from a file. If type is BUFFER, read from a memory buffer of CSV data */
  size_t	buffer_size;		/**< If type is BUFFER, the size of the data, unused otherwise */
```
 Or alternatively, a union.

Reply:
I think it would be fine replicating two fields, especially since it's doubtful to re-use structs initialized with the same values across different APIs. If we need to refactor, we can refactor.